libsignal-service/src/sender.rs

```diff
@@ -192,157 +208,264 @@ where
+            .try_send_message(
+                recipient.clone(),
+                unidentified_access,
+                &content_body,
+                timestamp,
+                online,
+            )
+            .await;
+
+        match result {
+            Ok(SendMessageResult { needs_sync, .. }) if needs_sync => {
+                log::debug!("sending multi-device sync message");
+                let sync_message = self
+                    .create_multi_device_sent_transcript_content(
+                        Some(&recipient),
+                        Some(message),
+                        timestamp,
+                    )?;
+                self.try_send_message(
+                    (&self.cipher.local_address).clone(),
+                    None,
+                    &sync_message,
                     timestamp,
-                    online,
+                    false,
                 )
-                .await
-            {
-                Err(MessageSenderError::TryAgain) => continue,
-                r => return r,
+                .await?;
             }
+            _ => (),
         }
-        Err(MessageSenderError::MaximumRetriesLimitExceeded)
+
+        if end_session {
+            log::debug!("ending session with {}", recipient);
+            if let Some(ref uuid) = recipient.uuid {
+                self.cipher.store_context.delete_all_sessions(&uuid)?;
+            }
+            if let Some(ref e164) = recipient.e164 {
+                self.cipher.store_context.delete_all_sessions(&e164)?;
+            }
+        }
+
+        result
     }
 
-    async fn send_messages(
+    /// Send a message to the recipients in a group.
+    pub async fn send_message_to_group(
         &mut self,
-        recipient: &ServiceAddress,
-        _unidentified_access: Option<&UnidentifiedAccess>,
-        content: &[u8],
+        recipients: Vec<ServiceAddress>,
+        unidentified_access: Option<&UnidentifiedAccess>,
+        message: crate::proto::DataMessage,
         timestamp: u64,
         online: bool,
-    ) -> Result<SendMessageResponse, MessageSenderError> {
-        let messages = self
-            .create_encrypted_messages(&recipient, None, &content)
-            .await?;
-
-        let messages = OutgoingPushMessages {
-            destination: recipient.identifier(),
-            timestamp,
-            messages,
-            online,
-        };
+    ) -> Result<
+        Vec<Result<SendMessageResult, MessageSenderError>>,
+        MessageSenderError,
+    > {
+        // TODO: discuss and fix the rather horrible type above
+        let content_body: ContentBody = message.clone().into();
+
+        let mut send_futures = vec![];
+        for recipient in recipients.iter() {
+            send_futures.push(self.try_send_message(
+                recipient.clone(),
+                unidentified_access,
+                &content_body,
+                timestamp,
+                online,
+            ))
+        }
```

Comment:
No need to allocate, I think. Can't you `let send_futures = recipient.iter().map(|recipient| self.try_send_message(...))`? `join_all` down below accepts iterators, then you don't have to allocate twice.